Update intro and plotting (#8)
duke-malaria-collaboratory/DiscovR: _episodes_rmd/02-r-plotting.Rmd

> teaching: 90

Reviewer: Check to be sure these line up with the times on the schedule.

gapminder_1997 <- read_csv("data/gapminder_1997.csv")
```

# Loading and reviewing data

> ## The tidyverse vs Base R
> If you've used R before, you may have learned commands that are different than the ones we will be using during this workshop. We will be focusing on functions from the [tidyverse](https://www.tidyverse.org/). The "tidyverse" is a collection of R packages that have been designed to work well together and offer many convenient features that do not come with a fresh install of R (aka "base R"). These packages are very popular and have a lot of developer support including many staff members from RStudio. These functions generally help you to write code that is easier to read and maintain. We believe learning these tools will help you become more productive more quickly.
{: .callout}

> First, we're going to load the tidyverse, which contains a set of useful functions that makes it easier for us to do complex analyses and create professional visualizations in R. The way we access alk these useful functions functions is by running the following command:

Reviewer: Can we expand this a little bit to also explain what a package is? Here is what I'm thinking (feel free to edit as needed and note I also fixed a typo alk -> all):

First, we're going to load the tidyverse package. Packages are useful because they contain pre-made functions to do specific tasks. Tidyverse contains a set of functions that makes it easier for us to do complex analyses and create professional visualizations in R. The way we access all of these useful functions functions is by running the following command:

> you'll be introduced to tomorrow in the 
> [R for Data Analysis lesson]({{ page.root }}/05-r-markdown)).
>
> The second section of messages gives a list of "conflicts." Sometimes, the
> same function name will be used in two different packages, and R has to decide
> which function to use. For example, our message says that:
>
> ~~~
> dplyr::filter() masks stats::filter()
> ~~~
> {: .output}
>
> This means that two different packages (`dyplr` from `tidyverse` and `stats`
> from base R) have a function named `filter()`. By default, R uses the function
> that was most recently loaded, so if we try using the `filter()` function after
> loading `tidyverse`, we will be using the `filter()` function from `dplyr()`.
>
{: .callout}

Okay, now let's read in our data, gapminder_1997.csv, which is in the data folder in our project directory. 
We're going to use the `read_csv()` function that we loaded in with the tidyverse, 
and save it to `gapminder_1997`, which will act as a placeholder for our data.
This function takes a file path and returns a tibble, which is basically a table (that we sometimes call a data frame...).

```{r consoleReadEcho, eval = FALSE}

> gapminder_1997 <- read_csv("gapminder_1997.csv")

Reviewer: Did you mean to have this here?

```
```{r consoleReadEval, echo = FALSE}
gapminder_1997 <- read_csv("data/gapminder_1997.csv")
```

First, a few things printed out to the screen: it tells us how many rows and columns are in our data, 
and information about each of the columns. 
Each row contains life expectancy ("lifeExp"), the total population ("pop"), and the per capita gross domestic product ("gdpPercap") for a given country ("country").
There is also a column that says which continent each country is in ("continent"). Note that both North America and South America are combined into one category called "Americas".
We can see that two of the columns are characters (categorical variables), and three are doubles (numbers).

> **Note** In anything before R 4.0, categorical variables used to be read in as factors, which are a [special data object](https://www.tutorialspoint.com/r/r_factors.htm) that are used to store categorical data and have limited numbers of unique values. The unique values of a factor are tracked via the "levels" of a factor. A factor will always remember all of its levels even if the values don't actually appear in your data. The factor will also remember the order of the levels and will always print values out in the same order (by default this order is alphabetical).

Reviewer: little comment: "factors, which are [special data object**s**]"

Author: ```suggestion
**Note** In anything before R 4.0, categorical variables used to be read in as factors, which are [special data objects](https://www.tutorialspoint.com/r/r_factors.htm) that are used to store categorical data and have limited numbers of unique values. The unique values of a factor are tracked via the "levels" of a factor. A factor will always remember all of its levels even if the values don't actually appear in your data. The factor will also remember the order of the levels and will always print values out in the same order (by default this order is alphabetical).
```

> ## Data objects
> There are many different ways to store data in R. Most objects have a table-like structure with rows and columns. We will refer to these objects generally as "data objects". If you've used R before, you many be used to calling them "data.frames". Functions from the "tidyverse" such as `read_csv()` work with objects called "tibbles", which are a specialized kind of "data.frame." Another common way to store data is a "data.table". All of these types of data objects (tibbles, data.frames, and data.tables) can be used with the commands we will learn in this lesson to make plots. We may sometimes use these terms interchangeably.
{: .callout}

> ## Bonus: Reading in an excel file
> Say you have an excel file and not a csv - how would you read that in? Hint: Use the Internet to help you figure it out!
>
> {: .source}
>
> > ## Solution
> > One way is using the `read_excel` function in the `readxl` package. There are other ways, but this is our preferred method because the output will be the same as the output of `read_csv`.
> {: .solution}
{: .challenge}

# Our first plot
_[Back to top](#contents)_

## Creating our first plot
_[Back to top](#contents)_

> We will be using the `ggplot2` package today to make our plots. This is a very

Reviewer: As noted above, I think we need to be careful in how we introduce aesthetics (lines 178 - 181). I think some more clear language would be:

The elements of a plot have a bunch of properties like an x and y position, a size, a color, etc. When creating a data visualization, we can map variables in our dataset to these properties, called **aesthetics**, in our plot. In ggplot, we can do this by creating an "aesthetic mapping", which we do with the `aes()` function.

Might want to bold "geometry" on line 253.

Another way to look at the data is to print it out to the console:

```{r, viewGapminder}
gapminder_1997
```

The `read_csv()` function took the file path we provided, did who-knows-what behind the scenes, and then outputted an R object with the data stored in that csv file. All that, with one short line of code!

> ## Data objects
> There are many different ways to store data in R. Most objects have a table-like structure with rows and columns. We will refer to these objects generally as "data objects". If you've used R before, you many be used to calling them "data.frames". Functions from the "tidyverse" such as `read_csv()` work with objects called "tibbles", which are a specialized kind of "data.frame." Another common way to store data is a "data.table". All of these types of data objects (tibbles, data.frames, and data.tables) can be used with the commands we will learn in this lesson to make plots. We may sometimes use these terms interchangeably.
{: .callout}

> ## Bonus: Reading in an excel file
> Say you have an excel file and not a csv - how would you read that in? Hint: Use the Internet to help you figure it out!
>
> {: .source}
>
> > ## Solution

> > > One way is using the `read_excel` function in the `readxl` package. There are other ways, but this is our preferred method because the output will be the same as the output of `read_csv`.

Reviewer: ```suggestion
> > One way is using the `read_excel` function in the `readxl` package. Hint: you may need to use `install.packages()` to install the `readxl` package. There are other ways to read in excel files, but this is our preferred method because the output will be the same as the output of `read_csv`.
```

> {: .solution}
{: .challenge}

# Our first plot

## Creating our first plot
_[Back to top](#contents)_

> We will be using the `ggplot2` package today to make our plots. This is a very

Reviewer: ```suggestion
We will be using the `ggplot2` package, which is part of `tidyverse`, to make our plots. This is a very
```

    title = "Do people in wealthy countries live longer?", size = "Population (in millions)")
```

> ## Understanding common errors
> Sometimes you accidentally type things wrong and get errors. We call these mis-types "bugs".
> Let's go through some common ones. The most important things to remember are:
> 1. The order of parentheses, quotes, commas, and plusses matters.
> 1. Sometimes you accidentally forget a plus where you need one or include one where you don't.
>
> For each of the examples below, figure out what the bug is and how to fix it. 
> Feel free to copy/paste into RStudio to help you figure it out.
> 
> ```{r plotErrors, eval = FALSE}
> # Bug 1
> 
> ```
> 
> {: .source}
> > ## Solution
> > ```{r, eval = FALSE}
> >  gdp_vs_life <- ggplot(data = gapminder_1997) +
> >  aes(x = gdpPercap, y = lifeExp, color = continent, size = pop/1000000) +
> >  geom_point() +
> >  scale_color_brewer(palette = "Set1") +
> >  labs(x = "GDP Per Capita", y = "Life Expectancy",
> >    title = "Do people in wealthy countries live longer?", size = "Population (in millions)")
> > ```
> > {: .source}
> {: .solution}
{: .challenge}

# Plotting for data exploration
_[Back to top](#contents)_

> 

Reviewer: On line 564, we mention the interactive RStudio file navigation, which we got rid of. Can remove that bit. I also can't tell if the exercise answer is already provided just above it? This is an un-commentable bit for me, so it's hard to tell.

```{r}
base_plot +
  aes(fill = continent) +
  geom_boxplot()
```

Let's say we want to change the fill of our plots, but to all the same color. Maybe we want our boxplots to be "cadetblue". Let's try it out:

```{r GapViolinFillWrong}
base_plot +
  aes(fill="cadetblue") +
  geom_boxplot()
```
Hmm that's not quite what we want. In this example, we placed the fill inside the `aes()` function. Because we are using an aesthetic mapping, the "scale" for the fill will assign colors to values - in this case, we only have one value: the word "cadetblue." Instead, we can do this by explicitly setting the color aesthetic inside the `geom_boxplot` function. Because we are assigning a color directly and not using any values from our data to do so, we do not need to use the `aes()` mapping function. Let's try it out:

```{r GapViolinFill}
base_plot +
  geom_boxplot(fill="cadetblue")
```

That's better! R knows lots of color names. You can see the full list if you run `colors()` in the console. Since there are so many, you can randomly choose 10 if you run `sample(colors(), size = 10)`.

> ## Bonus: Choosing a color

> >  Use `sample(colors(), size = 10)` a few times until you get an interesting sounding color name and swap that out for "cadetblue" in the violin plot example.

Reviewer: Change violin plot to boxplot.

```{r GapViolinJitterLayers}
base_plot +
  aes(color = continent) +
  geom_boxplot() +
  geom_point(alpha = 0.3)
```

So how do we make it so that just the points are colored but not the boxplots? Each layer can have it's own set of aesthetic mappings. So far we've been using `aes()` outside of the other functions. When we do this, we are setting the "default" aesthetic mappings for the plot. But we can also set the asethetics inside the specific geom that we want to change. To do that, you can place an additional `aes()` inside of that layer:

```{r GapViolinJitterSpecific}
base_plot +
  geom_boxplot() +
  geom_point(aes(color = continent), alpha = 0.3)
```

Nice! Both `geom_boxplot` and `geom_point` will inherit the default values of `aes(continent, lifeExp)` in the base plot, but only `geom_point` will also use `aes(color = continent)`.

> ## Aesthetics inside the `ggplot()` function
> We can also map our default aesthetics by passing the values to the `ggplot()` function call as is sometimes more common:
> ```{r}
> ggplot(data = gapminder_1997, mapping = aes(x = continent, y = lifeExp)) +
>   geom_boxplot() +
>   geom_point()
> ```
{: .callout}

## Univariate Plots

> _[Back to top](#contents)_

Reviewer: Add bar charts here - univariate discrete.

> {: .solution}
{: .challenge}

### Univariate discrete

What if we want to plot a univariate discrete variable, like continent? For this, we can use a bar chart. 

> ## Exercise: Discrete univariate plots

> > Create a bar plot of `continent`. You can try guessing the geom or look it up on the cheat sheet or Internet. 

Reviewer: ```suggestion
> Create a bar plot of `continent` that shows the number of countries in each continent. You can try guessing the geom or look it up on the cheat sheet or Internet. 
```

> # Glossary of terms

Reviewer: Above: applying to your own data, we will want to make the questions/assignment more concrete and also print out worksheets for folks to print out.

Author: Will address this later
